File: dashboard/src/shared/analytics/dayShape.ts
import type { DayJournal } from "@shared/api";

export interface HourLoad {
  hour: number;
  minutes: number;
  level: 0 | 1 | 2 | 3 | 4;
}

export interface DayShape {
  hours: HourLoad[];
  activeMinutes: number;
  startedAt: string | null;
  endedAt: string | null;
  peakHour: number | null;
}

function hourLevel(minutes: number): 0 | 1 | 2 | 3 | 4 {
  if (minutes < 1) return 0;
  if (minutes <= 15) return 1;
  if (minutes <= 30) return 2;
  if (minutes <= 45) return 3;
  return 4;
}

export function dayShape(journal: DayJournal | null): DayShape {
  const minutes = Array.from({ length: 24 }, () => 0);
  const empty: DayShape = {
    hours: minutes.map((_, hour) => ({ hour, minutes: 0, level: 0 as const })),
    activeMinutes: 0,
    startedAt: null,
    endedAt: null,
    peakHour: null,
  };
  if (!journal) return empty;

  const midnight = new Date(`${journal.date}T00:00:00`);
  if (Number.isNaN(midnight.valueOf())) return empty;
  const bounds = Array.from({ length: 25 }, (_, hour) => {
    const edge = new Date(midnight);
    edge.setHours(hour, 0, 0, 0);
    return edge.valueOf();
  });

  for (const session of journal.sessions ?? []) {
    const active = Math.max(0, session.activeMinutes ?? 0);
    if (active <= 0) continue;
    const start = Date.parse(session.startedAt);
    if (!Number.isFinite(start)) continue;
    const parsedEnd = Date.parse(session.endedAt);
    const end = Number.isFinite(parsedEnd) && parsedEnd > start ? parsedEnd : start;
    const overlaps = bounds.slice(0, 24).map((from, hour) => {
      const to = bounds[hour + 1];
      return Math.max(0, Math.min(end, to) - Math.max(start, from));
    });
    const covered = overlaps.reduce((sum, span) => sum + span, 0);
    if (covered <= 0) {
      const hour = Math.min(23, Math.max(0, new Date(start).getHours()));
      minutes[hour] += active;
      continue;
    }
    for (let hour = 0; hour < 24; hour++) {
      if (overlaps[hour] > 0) minutes[hour] += active * (overlaps[hour] / covered);
    }
  }

  const hours = minutes.map((value, hour) => {
    const capped = Math.min(60, value);
    return { hour, minutes: capped, level: hourLevel(capped) };
  });
  const busiest = hours.reduce((best, entry) => (entry.minutes > best.minutes ? entry : best), hours[0]);
  return {
    hours,
    activeMinutes: journal.activeMinutes ?? 0,
    startedAt: journal.firstEventAt,
    endedAt: journal.lastEventAt,
    peakHour: busiest.minutes > 0 ? busiest.hour : null,
  };
}

export function formatClock(iso: string | null, locale?: string): string {
  if (!iso) return "—";
  const date = new Date(iso);
  if (Number.isNaN(date.valueOf())) return "—";
  return new Intl.DateTimeFormat(locale, { hour: "numeric", minute: "2-digit" }).format(date);
}

export function formatHourLabel(hour: number, locale?: string): string {
  const date = new Date(2000, 0, 1, hour);
  return new Intl.DateTimeFormat(locale, { hour: "numeric" }).format(date);
}

File: macos/Sources/ObolCore/Usage/DayShape.swift
import Foundation

/// The 24-hour activity projection shown by the menu-bar popover.
///
/// This is deliberately Foundation-only: the view owns the color ramp and
/// layout, while this type owns the time bucketing and ISO-8601 parsing.
public struct DayShape: Equatable, Sendable {
    public let activeMinutes: Double
    public let levels: [Int]
    public let startedAt: String?
    public let peakHour: Int?

    public init(
        activeMinutes: Double = 0,
        levels: [Int] = Array(repeating: 0, count: 24),
        startedAt: String? = nil,
        peakHour: Int? = nil
    ) {
        self.activeMinutes = activeMinutes
        self.levels = levels
        self.startedAt = startedAt
        self.peakHour = peakHour
    }

    /// Distributes each session's active minutes across the wall-clock hours
    /// it covered, preserving the popover's original weighting semantics.
    public static func from(_ journal: DayJournal, calendar: Calendar = .current) -> DayShape {
        var minutes = Array(repeating: 0.0, count: 24)
        guard let first = parseDate(journal.firstEventAt),
              let midnight = calendar.date(
                  from: calendar.dateComponents([.year, .month, .day], from: first)
              )
        else {
            return DayShape()
        }

        let bounds = (0 ... 24).compactMap {
            calendar.date(byAdding: .hour, value: $0, to: midnight)
        }
        guard bounds.count == 25 else { return DayShape() }

        for session in journal.sessions {
            guard let start = parseDate(session.startedAt) else { continue }
            let end = parseDate(session.endedAt) ?? start
            let spans = (0 ..< 24).map { hour in
                max(
                    0,
                    min(end.timeIntervalSince1970, bounds[hour + 1].timeIntervalSince1970)
                        - max(start.timeIntervalSince1970, bounds[hour].timeIntervalSince1970)
                )
            }
            let covered = spans.reduce(0, +)
            if covered > 0 {
                for hour in 0 ..< 24 {
                    minutes[hour] += session.activeMinutes * spans[hour] / covered
                }
            }
        }

        let peak = minutes.enumerated().max(by: { $0.element < $1.element })
        return DayShape(
            activeMinutes: journal.activeMinutes,
            levels: minutes.map(level),
            startedAt: journal.firstEventAt,
            peakHour: peak?.offset
        )
    }

    public static func level(_ minutes: Double) -> Int {
        minutes < 1 ? 0 : minutes <= 15 ? 1 : minutes <= 30 ? 2 : minutes <= 45 ? 3 : 4
    }

    public static func duration(_ minutes: Double) -> String {
        "\(Int(minutes) / 60)h \(Int(minutes) % 60)m"
    }

    public static func hourLabel(_ hour: Int) -> String {
        hour == 0 ? "12" : hour > 12 ? "\(hour - 12)" : "\(hour)"
    }

    public static func clock(_ iso: String?) -> String {
        guard let date = parseDate(iso) else { return "—" }
        return date.formatted(.dateTime.hour().minute())
    }

    public static func parseDate(_ iso: String?) -> Date? {
        guard let iso else { return nil }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.date(from: iso) ?? {
            formatter.formatOptions = [.withInternetDateTime]
            return formatter.date(from: iso)
        }()
    }

    public static func accessibilityLabel(_ shape: DayShape) -> String {
        "Work started at \(clock(shape.startedAt)), \(duration(shape.activeMinutes)) active today"
            + (shape.peakHour.map { ", busiest around \(hourLabel($0))" } ?? "")
    }
}
